File: server/models/characters/ghost.py
from models.interfaces import Direction, CellType
from pacman_constants import DIRECTION_MAP
from models.map.game_map import GameMap
from typing import Tuple
from enum import Enum

import numpy as np

class GhostType(Enum):
    BLINKY = 0
    PINKY = 1
    CLYDE = 2
    INKY = 3

class GhostState(Enum):
    FRIGHTENED = 0
    SCATTER = 1
    CHASE = 2

class Ghost:
    def __init__(self, ghost_type: GhostType, x: int, y: int, game_map: GameMap, cell_size: int):
        self.state = GhostState.SCATTER
        self.type = ghost_type
        
        self.cell_size = cell_size
        self.is_game_over = False
        self.game_map = game_map
        
        self.gridX = x
        self.gridY = y
        
        self.displayX = x * cell_size
        self.displayY = y * cell_size
        
        self.next_direction = Direction(0, 0)
        self.direction = Direction(0, 0)
        self.is_moving = False
        
        self.move_interval = 225 * 1.3
        self.last_state_change = 0
        self.last_move_time = 0
        
        self.scatter_target = self._get_scatter_target()
        self.exited_room = False
        self.cycle_index = 0
    
    def update(self, current_time: int, game_state: dict):
        if self.is_game_over:
            return
        
        if not self._update_position(current_time):
            return
        
        self._handle_state_transition(current_time)
        self._update_speed(current_time)
        self._check_door_exit()
        
        self._determine_move_direction(game_state)
        
        self._move_ghost(current_time)
    
    def _update_position(self, current_time: int) -> bool:
        if self.is_game_over:
            return False
        
        targetX = self.gridX * self.cell_size
        targetY = self.gridY * self.cell_size
        
        self.is_moving = (
            abs(self.displayX - targetX) > 0.1 or 
            abs(self.displayY - targetY) > 0.1 or 
            (self.direction.x != 0 or self.direction.y != 0)
        )
        
        if current_time - self.last_move_time < self.move_interval:
            self.displayX += (targetX - self.displayX) * 0.1
            self.displayY += (targetY - self.displayY) * 0.1
            return False
            
        return True
    
    def _check_door_exit(self):
        if self.game_map.get_cell(self.gridX, self.gridY) == CellType.Door:
            next_cell = self.game_map.get_cell(
                self.gridX + self.direction.x, 
                self.gridY + self.direction.y
            )
            self.exited_room = next_cell != CellType.GhostCell
    
    def _determine_move_direction(self, game_state: dict):
        if self.state == GhostState.SCATTER:
            self.direction = self._get_direction_towards(game_state, self.scatter_target)
        elif self.state == GhostState.CHASE:
            self.direction = self._select_chase_direction(game_state)
        else:  # FRIGHTENED
            self.direction = self._get_valid_random_direction()
        
        # If direction is invalid, choose a random valid one
        if not self._is_valid_position(self.gridX + self.direction.x, self.gridY + self.direction.y):
            self.direction = self._get_valid_random_direction()
    
    def _is_valid_position(self, x: int, y: int) -> bool:
        if not (0 <= x < self.game_map.width and 0 <= y < self.game_map.height):
            return False
        
        if not self.game_map.is_walkable(x, y, self._can_pass_door()):
            return False
        
        valid_directions = [
            d for d in DIRECTION_MAP 
            if self.game_map.is_walkable(self.gridX + d.x, self.gridY + d.y, self._can_pass_door())
        ]
        
        # If only one valid direction, allow it even if it's going back
        if len(valid_directions) == 1:
            return True
        
        # Prevent turning around
        is_opposite_direction = (
            x == self.gridX - self.direction.x and 
            y == self.gridY - self.direction.y
        )
        
        return not is_opposite_direction
    
    def _can_pass_door(self) -> bool:
        return not self.exited_room
    
    def _move_ghost(self, current_time: int):
        newX = self.gridX + self.direction.x
        newY = self.gridY + self.direction.y
        
        if self._is_valid_position(newX, newY):
            self.last_move_time = current_time
            self.gridX = newX
            self.gridY = newY
    
    def _handle_state_transition(self, current_time: int):
        scatter_chase_cycle = [7000, 20000, 7000, 20000, 5000, 20000, 5000, float('inf')]
        since_last_change = current_time - self.last_state_change
        
        if self.state == GhostState.FRIGHTENED and since_last_change > 7000:
            self.state = GhostState.CHASE
            self.last_state_change = current_time
            return
        
        if since_last_change > scatter_chase_cycle[self.cycle_index]:
            self.state = GhostState.SCATTER if self.state == GhostState.CHASE else GhostState.CHASE
            self.cycle_index = min(self.cycle_index + 1, len(scatter_chase_cycle) - 1)
            self.last_state_change = current_time
    
    def _update_speed(self, current_time: int):
        if self.state == GhostState.FRIGHTENED:
            self.move_interval = 225 * 1.3 * 0.8  # Slower when frightened
            return
        
        # Increase speed over time
        speed_increase_interval = 10000
        speed_increase = min(np.floor(current_time / speed_increase_interval) * 0.05, 0.15)
        self.move_interval = 225 * max(1.3 - speed_increase, 1.15)
    
    def _get_scatter_target(self) -> Tuple[int, int]:
        scatter_targets = {
            GhostType.INKY: (self.game_map.width - 1, self.game_map.height - 1),
            GhostType.BLINKY: (self.game_map.width - 1, 0),
            GhostType.CLYDE: (0, self.game_map.height - 1),
            GhostType.PINKY: (0, 0)
        }
        return scatter_targets[self.type]
    
    def _get_direction_towards(self, game_state: dict, target: Tuple[int, int]) -> Direction:
        width, height = self.game_map.width, self.game_map.height
        targetX, targetY = target
        
        targetX = max(0, min(targetX, width - 1))
        targetY = max(0, min(targetY, height - 1))
        
        queue = [(self.gridX, self.gridY)]
        cost = [[None for _ in range(height)] for _ in range(width)]
        cost[self.gridX][self.gridY] = {'distance': 0, 'parent': None}
        
        closest = {
            'x': self.gridX,
            'y': self.gridY,
            'distance': abs(self.gridX - targetX) + abs(self.gridY - targetY)
        }
        
        while queue:
            x, y = queue.pop(0)
            distance = abs(x - targetX) + abs(y - targetY)
            
            if distance < closest['distance']:
                closest = {'x': x, 'y': y, 'distance': distance}
            
            if x == targetX and y == targetY:
                break
            
            for direction in DIRECTION_MAP:
                nx, ny = x + direction.x, y + direction.y
                
                if not self._is_valid_position(nx, ny) or cost[nx][ny] is not None:
                    continue
                
                cost[nx][ny] = {
                    'distance': cost[x][y]['distance'] + 1,
                    'parent': (-direction.x, -direction.y)
                }
                queue.append((nx, ny))
        
        return self._trace_back_direction(cost, closest['x'], closest['y'])
    
    def _trace_back_direction(self, cost, x: int, y: int) -> Direction:
        while cost[x][y] and cost[x][y]['parent']:
            parent_dx, parent_dy = cost[x][y]['parent']
            px, py = x + parent_dx, y + parent_dy
            
            if px == self.gridX and py == self.gridY:
                return Direction(x - px, y - py)
            
            x, y = px, py
        
        return self._get_valid_random_direction()
    
    def _get_valid_random_direction(self) -> Direction:
        valid_directions = [
            direction for direction in DIRECTION_MAP
            if self._is_valid_position(self.gridX + direction.x, self.gridY + direction.y)
        ]
        
        if valid_directions:
            return valid_directions[np.random.choice(len(valid_directions))]
        return self.direction
    
    def _select_chase_direction(self, game_state: dict) -> Direction:
        return self._get_valid_random_direction()
    
    def enter_frightened_state(self, current_time: int):
        self.state = GhostState.FRIGHTENED
        self.last_state_change = current_time
    
    def on_eaten(self, current_time: int):
        self.last_state_change = current_time
        self.state = GhostState.SCATTER
        
        respawn_points = {
            GhostType.BLINKY: (9, 9),
            GhostType.PINKY: (9, 9),
            GhostType.INKY: (10, 9),
            GhostType.CLYDE: (8, 9)
        }
        
        self.gridX, self.gridY = respawn_points[self.type]
        self.exited_room = False
    
    def is_frightened(self) -> bool:
        return self.state == GhostState.FRIGHTENED
    
    def set_game_over(self):
        self.is_game_over = True
        self.next_direction = Direction(0, 0)


class Blinky(Ghost):
    def __init__(self, x: int, y: int, game_map: GameMap, cell_size: int):
        super().__init__(GhostType.BLINKY, x, y, game_map, cell_size)
    
    def _select_chase_direction(self, game_state: dict) -> Direction:
        pacman_pos = game_state['pacman_position']
        return self._get_direction_towards(game_state, (pacman_pos['x'], pacman_pos['y']))


class Pinky(Ghost):
    def __init__(self, x: int, y: int, game_map: GameMap, cell_size: int):
        super().__init__(GhostType.PINKY, x, y, game_map, cell_size)
    
    def _select_chase_direction(self, game_state: dict) -> Direction:
        pacman_pos = game_state['pacman_position']
        return self._get_direction_towards(game_state, (pacman_pos['x'] + 4, pacman_pos['y']))


class Inky(Ghost):
    def __init__(self, x: int, y: int, game_map: GameMap, cell_size: int):
        super().__init__(GhostType.INKY, x, y, game_map, cell_size)
    
    def _select_chase_direction(self, game_state: dict) -> Direction:
        pacman_pos = game_state['pacman_position']
        ghost_positions = game_state['ghost_positions']
        
        blinky = next((g for g in ghost_positions if g['type'] == GhostType.BLINKY), None)
        if blinky:
            target = (
                pacman_pos['x'] + (pacman_pos['x'] - blinky['x']),
                pacman_pos['y'] + (pacman_pos['y'] - blinky['y'])
            )
            return self._get_direction_towards(game_state, target)
        
        return self._get_valid_random_direction()


class Clyde(Ghost):
    def __init__(self, x: int, y: int, game_map: GameMap, cell_size: int):
        super().__init__(GhostType.CLYDE, x, y, game_map, cell_size)
    
    def _select_chase_direction(self, game_state: dict) -> Direction:
        pacman_pos = game_state['pacman_position']
        
        distance = np.sqrt(
            (self.gridX - pacman_pos['x'])**2 + 
            (self.gridY - pacman_pos['y'])**2
        )
        
        if distance > 8:
            return self._get_direction_towards(game_state, (pacman_pos['x'], pacman_pos['y']))
        else:
            return self._get_direction_towards(game_state, self.scatter_target)

File: client/src/app/models/characters/ghost.ts
import {Character, POSSIBLE_DIRECTIONS, MOVE_INTERVAL} from './character';
import {Direction, GameState} from '@models/interfaces';
import {CellType} from '@models/map/cell-type';
import {GameMap} from '@models/map/game-map';
import {GhostState} from './ghost-state';
import {GhostType} from './ghost-type';

const GHOST_COLORS: Record<GhostType, string> = {
    [GhostType.BLINKY]: "#FF0000", // Blinky -> Red
    [GhostType.PINKY]: "#EB94BC",  // Pinky  -> Pink
    [GhostType.INKY]: "#00FCFF",   // Inky   -> Blue
    [GhostType.CLYDE]: "#FDCC08"   // Clyde  -> Orange
};

const GHOST_RESPAWN_POINT: Record<GhostType, Direction> = {
    [GhostType.BLINKY]: {x: 9, y: 9},
    [GhostType.PINKY]: {x: 9, y: 9},
    [GhostType.INKY]: {x: 10, y: 9},
    [GhostType.CLYDE]: {x: 8, y: 9}
};

const GHOST_RESPAWN_TIME: Record<GhostType, number> = {
    [GhostType.BLINKY]: 1500,
    [GhostType.PINKY]:  2000,
    [GhostType.INKY]:   4000,
    [GhostType.CLYDE]:  6000
};

const SCATTER_CHASE_CYCLE = [7000, 20000, 7000, 20000, 5000, 20000, 5000, Infinity];

const GHOST_BASE_SPEED = MOVE_INTERVAL * 1.3;
const SPEED_INCREASE_INTERVAL = 10000;

type NodePath = {
    distance: number;
    parent: Direction | null;
} | null;

export abstract class Ghost extends Character {    
    protected scatterTarget: Direction;
    protected previousState: GhostState;

    protected lastStateChange = 0;
    protected exitedRoom = false;
    protected cycleIndex = 0;
    
    state = GhostState.SCATTER;
    readonly type: GhostType;

    constructor(type: GhostType, x: number, y: number, gameMap: GameMap, cellSize: number, simulationMode: boolean) {
        super(gameMap, x, y, cellSize, GHOST_BASE_SPEED, simulationMode);
        this.type = type;

        this.scatterTarget = this.getScatterTarget(type, gameMap);
        this.previousState = GhostState.SCATTER;
    }

    update(currentTime: number, gameState: GameState): void {
        if (this.simulationMode) {
            const ghostData = gameState.ghostsInfo.find(g => g.type === this.type);
            if (ghostData) {
                super.updateFrom(
                    ghostData.x, 
                    ghostData.y, 
                    {x: ghostData.x - this.gridX, y: ghostData.y - this.gridY}, 
                    currentTime
                );
                this.state = ghostData.state;
            }

            return;
        }

        if (!this.updatePosition(currentTime)) return;

        this.handleStateTransition(currentTime);
        this.updateSpeed(currentTime);

        if (this.gameMap.getCell(this.gridX, this.gridY) === CellType.Door) {
            const nextCell = this.gameMap.getCell(this.gridX + this.direction.x, this.gridY + this.direction.y);
            this.exitedRoom = nextCell !== CellType.GhostCell;
        }

        if (!this.simulationMode) {
            switch (this.state) {
                case GhostState.SCATTER:
                    this.direction = this.getDirectionTowards(gameState, this.scatterTarget);
                    break;
                case GhostState.CHASE:
                    this.direction = this.selectChaseDirection(gameState);
                    break;
                default:
                    this.direction = this.getValidRandomDirection();
                    break;
            }
        }

        if (!this.simulationMode && !this.isValidPosition(this.gridX + this.direction.x, this.gridY + this.direction.y)) {
            this.direction = this.getValidRandomDirection();
        }

        this.moveGhost(currentTime);
    }

    draw(ctx: CanvasRenderingContext2D, currentTime: number): void {
        ctx.save();
        ctx.translate(this.displayX + this.cellSize / 2, this.displayY + this.cellSize / 2);
        
        this.drawBody(ctx, currentTime);
        this.drawEyes(ctx);
        this.drawPupils(ctx);
        
        ctx.restore();
    }

    enterFrightenedState(currentTime: number): void {
        if (this.state !== GhostState.FRIGHTENED) {
            this.previousState = this.state;
            this.state = GhostState.FRIGHTENED;
            this.lastStateChange = currentTime;
        }
    }

    onEaten(currentTime: number): void {
        this.direction = {x: 0, y: 0};
        this.lastMoveTime = currentTime;
        
        this.gridX = GHOST_RESPAWN_POINT[this.type].x;
        this.gridY = GHOST_RESPAWN_POINT[this.type].y;
        
        this.state = this.previousState;
        this.lastStateChange = currentTime;
        this.exitedRoom = true;
        
        setTimeout(() => {
            this.exitedRoom = false;
        }, GHOST_RESPAWN_TIME[this.type]);
    }

    isFrightened(): boolean {
        return this.state == GhostState.FRIGHTENED;
    }
    
    override canPassDoor(): boolean {
        return !this.exitedRoom;
    }

    override isValidPosition(x: number, y: number): boolean {
        if (!super.isValidPosition(x, y)) return false;
    
        const validMoves = POSSIBLE_DIRECTIONS.filter(({ x: dx, y: dy }) =>
            super.isValidPosition(this.gridX + dx, this.gridY + dy)
        );
    
        if (validMoves.length === 1) return true;
    
        const isOppositeDirection = x === this.gridX - this.direction.x && y === this.gridY - this.direction.y;
        return !isOppositeDirection;
    }

    protected getDirectionTowards(gameState: GameState, target: Direction): Direction {
        const {width, height} = gameState.gameMap;
        
        const targetX = Math.max(0, Math.min(target.x, width - 1));
        const targetY = Math.max(0, Math.min(target.y, height - 1));
    
        let queue: [number, number][] = [[this.gridX, this.gridY]];
        let cost: NodePath[][] = Array.from({ length: width }, () =>
            Array.from({ length: height }, () => null)
        );
    
        cost[this.gridX][this.gridY] = {distance: 0, parent: null};
        let closest = {
            x: this.gridX,
            y: this.gridY, 
            distance: Math.abs(this.gridX - targetX) + Math.abs(this.gridY - targetY) 
        };
    
        while (queue.length) {
            const [x, y] = queue.shift()!;
            const distance = Math.abs(x - targetX) + Math.abs(y - targetY);
    
            if (distance < closest.distance) closest = {x, y, distance};
            if (x === targetX && y === targetY) break;
    
            for (const dir of POSSIBLE_DIRECTIONS) {
                const nx = x + dir.x, ny = y + dir.y;
    
                if (!this.isValidPosition(nx, ny) || cost[nx][ny]) continue;
    
                cost[nx][ny] = {
                    distance: cost[x][y]!.distance + 1, 
                    parent: { x: -dir.x, y: -dir.y } 
                };
                queue.push([nx, ny]);
            }
        }
    
        return this.traceBackDirection(cost, closest.x, closest.y);
    }

    protected getValidRandomDirection(): Direction {
        const validDirections = 
            POSSIBLE_DIRECTIONS.filter(({x, y}) => this.isValidPosition(this.gridX + x, this.gridY + y));
        
        const directionIndex = Math.floor(Math.random() * validDirections.length);
        return validDirections.length > 0 ? validDirections[directionIndex] : this.direction;    
    }

    protected abstract selectChaseDirection(gameState: GameState): Direction;

    private moveGhost(currentTime: number): void {
        const newX = this.gridX + this.direction.x;
        const newY = this.gridY + this.direction.y;

        if (this.isValidPosition(newX, newY)) {
            this.lastMoveTime = currentTime;

            this.gridX = newX;
            this.gridY = newY;
        }
    }

    private drawBody(ctx: CanvasRenderingContext2D, currentTime: number): void {
        let ghostColor = GHOST_COLORS[this.type];

        if (this.state === GhostState.FRIGHTENED) {
            const timeElapsed = currentTime - this.lastStateChange;    
            const isFlashing = timeElapsed >= 5500 && Math.floor((timeElapsed / 200) % 2) === 0;
            ghostColor = isFlashing ? "#fff3e0" : "#ff6f00";
        }

        ctx.beginPath();
        ctx.fillStyle = ghostColor;

        ctx.arc(0, -2, this.cellSize / 2 - 4, Math.PI, 0, false);
        
        ctx.rect(-this.cellSize / 2 + 4, -2, this.cellSize - 8, this.cellSize / 2 - 2);
        ctx.fill();
    }

    private drawEyes(ctx: CanvasRenderingContext2D): void {
        ctx.fillStyle = 'white';
        ctx.beginPath();
        ctx.arc(-5, -4, 4, 0, Math.PI * 2);
        ctx.arc(5, -4, 4, 0, Math.PI * 2);
        ctx.fill();
    }

    private drawPupils(ctx: CanvasRenderingContext2D): void {
        ctx.fillStyle = 'black';
        const lookDir = Math.atan2(this.direction.y, this.direction.x);
        const pupilX = Math.cos(lookDir) * 2;
        const pupilY = Math.sin(lookDir) * 2;
        
        ctx.beginPath();
        ctx.arc(-5 + pupilX, -4 + pupilY, 2, 0, Math.PI * 2);
        ctx.arc(5 + pupilX, -4 + pupilY, 2, 0, Math.PI * 2);
        ctx.fill();
    }

    handleStateTransition(currentTime: number): void {
        const sinceLastChange = currentTime - this.lastStateChange;
        if (this.state === GhostState.FRIGHTENED) {
            if (sinceLastChange > 7000) {
                this.lastStateChange = currentTime;
                this.state = this.previousState;
            }
            return;
        }
    
        if (sinceLastChange > SCATTER_CHASE_CYCLE[this.cycleIndex]) {
            this.state = this.state === GhostState.SCATTER ? GhostState.CHASE : GhostState.SCATTER;
            this.cycleIndex = Math.min(this.cycleIndex + 1, SCATTER_CHASE_CYCLE.length - 1);
            this.lastStateChange = currentTime;
        }
    }

    private updateSpeed(currentTime: number): void {
        if (this.state === GhostState.FRIGHTENED) {
            this.moveInterval = GHOST_BASE_SPEED * 0.8;
            return;
        }
        
        const speedIncrease = Math.min(Math.floor(currentTime / SPEED_INCREASE_INTERVAL) * 0.05);
        this.moveInterval = MOVE_INTERVAL * Math.max(1.3 - speedIncrease, 1.15);
    }

    private getScatterTarget(type: GhostType, gameMap: GameMap) {
        const SCATTER_TARGETS: Record<GhostType, Direction> = {
            [GhostType.INKY]: {x: gameMap.width - 1, y: gameMap.height - 1},
            [GhostType.BLINKY]: {x: gameMap.width - 1, y: 0},
            [GhostType.CLYDE]: {x: 0, y: gameMap.height - 1},
            [GhostType.PINKY]: {x: 0, y: 0}
        };

        return SCATTER_TARGETS[type];
    }

    private traceBackDirection(cost: NodePath[][], x: number, y: number): Direction {
        while (cost[x][y]?.parent) {
            const parent = cost[x][y]!.parent!;
            const px = x + parent.x, py = y + parent.y;
    
            if (px === this.gridX && py === this.gridY) {
                return {
                    x: x - px,
                    y: y - py
                };
            }
    
            x = px;
            y = py;
        }
        return this.getValidRandomDirection();
    }
}
